controllers/clusterresources/opensearchegressrules_controller.go

```diff
@@ -0,0 +1,183 @@
+	l := log.FromContext(ctx)
+
+	rule := &clusterresourcesv1beta1.OpenSearchEgressRules{}
+	err := r.Client.Get(ctx, types.NamespacedName{
+		Namespace: req.Namespace,
+		Name:      req.Name,
+	}, rule)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			return models.ExitReconcile, nil
+		}
+
+		l.Error(err, "Unable to fetch OpenSearch Egress Rules resource")
+
+		return models.ReconcileRequeue, err
+	}
+
+	// It`s handling resource deletion
+	if rule.DeletionTimestamp != nil {
+		err = r.handleDelete(ctx, l, rule)
+		if err != nil {
+			return models.ReconcileRequeue, err
+		}
+
+		return models.ExitReconcile, nil
+	}
+
+	// It`s handling resource creation
+	if rule.Status.ID == "" {
+		err = r.handleCreate(ctx, l, rule)
+		if err != nil {
+			return models.ReconcileRequeue, nil
+		}
+
+		return models.ExitReconcile, nil
+	}
+
+	return models.ExitReconcile, nil
+}
+
+func (r *OpenSearchEgressRulesReconciler) handleCreate(ctx context.Context, l logr.Logger, rule *clusterresourcesv1beta1.OpenSearchEgressRules) error {
+	err := r.API.CreateOpenSearchEgressRules(rule)
+	if err != nil {
+		l.Error(err, "failed to create OpenSearch Egress Rule resource on Instaclustr")
+		r.EventRecorder.Eventf(rule, models.Warning, models.CreationFailed,
+			"Failed to create OpenSearch Egress Rule on Instaclustr. Reason: %v", err,
+		)
+
+		return err
+	}
+
+	patch := rule.NewPatch()
+
+	// This hack was added because the Instaclustr API returns an egress rule id with a null value, but it should be equal to the pattern {clusterId}~{source}~{bindingId}
+	// This code could be removed after the Instaclustr API team fixes this bug
+	if rule.Status.ID == "" {
+		rule.Status.ID = fmt.Sprintf("%s~%s~%s", rule.Spec.ClusterID, rule.Spec.Source, rule.Spec.OpenSearchBindingID)
+	}
+
+	err = r.Status().Patch(ctx, rule, patch)
+	if err != nil {
+		l.Error(err, "failed to patch OpenSearch Egress Rule status with its id")
+		r.EventRecorder.Eventf(rule, models.Warning, models.PatchFailed,
+			"Failed to patch OpenSearch Egress Rule with its id. Reason: %v", err,
+		)
+
+		return err
+	}
+
+	controllerutil.AddFinalizer(rule, models.DeletionFinalizer)
+	err = r.Patch(ctx, rule, patch)
+	if err != nil {
+		l.Error(err, "failed to patch OpenSearch Egress Rule with finalizer")
+		r.EventRecorder.Eventf(rule, models.Warning, models.PatchFailed,
+			"Failed to patch OpenSearch Egress Rule with finalizer. Reason: %v", err,
+		)
+
+		return err
+	}
+
+	l.Info("OpenSearch Egress Rule has been created")
+	r.EventRecorder.Event(rule, models.Normal, models.Created,
+		"OpenSearch Egress Rule has been created",
+	)
+
+	return nil
+}
+
+func (r *OpenSearchEgressRulesReconciler) handleDelete(ctx context.Context, logger logr.Logger, rule *clusterresourcesv1beta1.OpenSearchEgressRules) error {
+	err := r.API.DeleteOpenSearchEgressRule(rule.Status.ID)
+	if err != nil {
+		logger.Error(err, "failed to delete OpenSearch Egress Rule on Instaclustr")
```

Comment:
I think it's better to handle the NotFound error

Reply:
fixed